common/workunit/workunit.cpp

```diff
@@ -5044,6 +5100,262 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+                IPropertyTree &fileTree = itr->query();
+
+                StringBuffer fileNameWithPath(unzipDir);
+                fileNameWithPath.append(PATHSEPSTR).append(pathTail(fileTree.queryProp("@filename")));
+                if (checkFileExists(fileNameWithPath)) //Check if the ZAP report contains this file.
+                {
+                    fileTree.setProp("@ip", localIP);
+                    fileTree.setProp("@filename", fileNameWithPath);
+                }
+            }
+        }
+        void getWUAttributes(IWorkUnit *workunit)
+        {
+            wuid.set(workunit->queryWuid());
+            scope.set(workunit->queryWuScope());
+            workunit->getWorkunitDistributedAccessToken(distributedAccessToken);
+        }
+        IPTree *queryWUPTree()
+        {
+            return wuTree;
+        }
+        IPTree *querygraphProgressPTree()
+        {
+            return graphProgressTree;
+        }
+        void setUNZIPDir()
+        {   //Set a unique unzip folder inside the component's data folder
+            unzipDir.append(importDir).append(PATHSEPSTR).append(wuid.get());
+        }
+        int unzipZAPReport()
+        {
+            Owned<IFile> unzipFolder = createIFile(unzipDir);
+            if (!unzipFolder->exists())
+                unzipFolder->createDirectory();
+            else
+            {//This should not happen. Just in case.
+                Owned<IDirectoryIterator> iter = unzipFolder->directoryFiles(nullptr, false, false);
+                ForEach(*iter)
+                {
+                    OwnedIFile thisFile = createIFile(iter->query().queryFilename());
+                    if (thisFile->isFile() == foundYes)
+                        thisFile->remove();
+                }
+            }
+
+            CDateTime dt;
+            dt.setNow();
+            dt.getString(unzipDateTime);
+
+            //Unzip ZAP Report
+            StringBuffer zipCommand("unzip ");
+            if (!isEmptyString(zapReportFilePath))
+                zipCommand.append(zapReportFilePath);
+            zipCommand.appendf("%s -d %s", zapReportFileName.get(), unzipDir.str());
+            if (!isEmptyString(zapReportPassword))
+                zipCommand.appendf(" -P %s", zapReportPassword.get());
+            return (system(zipCommand.str()));
+        }
+        void buildPTreesFromZAPReport()
+        {
+            readWUXMLFileToPTree(wuTree, "ZAPReport_W*.xml", false);
+            wuTree->setProp("@scope", scope);
+            wuTree->setProp("@distributedAccessToken", distributedAccessToken.str());
+
+            StringBuffer localIP, unzipDirWithIP("//");
+            IpAddress ipaddr = queryHostIP();
+            ipaddr.getIpText(localIP);
+            unzipDirWithIP.append(localIP).append(unzipDir).append(PATHSEPSTR);
+
+            //update log entries in WU XML;
+            updateWUProcessLogAttrs(unzipDirWithIP, false);
+            updateWUProcessLogAttrs(unzipDirWithIP, true);
```

Comment:
@wangkx - don't think this is quite right, based on our discussion, the intention was to clear the existing Process entries, then use standard methods, i.e. IWorkUnit::addProcess(), to add in the new entries, rather than trying to patch up the existing IPT.
Previous comment: https://github.com/hpcc-systems/HPCC-Platform/pull/13258#discussion_r401600423

Reply:
From our discussion, I may pass 0 as PID when calling addProcess(). The Process Instance Name can be parsed from the log file name. Number of thor slaves? For new WU, the existing process entry is the place to store the number of slaves. Should I read from there? For legacy WU, I guess I have to use the findNumberOfThorSlavesFromSlaveLogs().